8301214

djelinski opened the pull request:
Please review this patch that reduces the socket timeout used in HandshakeTimeout test to its minimum value of 1 millisecond.

This change makes the test complete 10 seconds faster; before this change it took 5 seconds for the handshake to timeout, and the test attempts 2 handshakes.

The change also makes the test more likely to pass when it has to compete with other tests for CPU time.

vyommani commented on test/jdk/java/rmi/transport/handshakeTimeout/HandshakeTimeout.java:
```diff
@@ -56,8 +56,7 @@ public class HandshakeTimeout {
 
-        System.setProperty("sun.rmi.transport.tcp.handshakeTimeout",
-                           String.valueOf(TIMEOUT / 2));
+        System.setProperty("sun.rmi.transport.tcp.handshakeTimeout", "1");
```
I can see that this test uses "TIMEOUT" down in test 
/*
         * Wait for call attempt to finished, and analyze result.
         */
        t.join(TIMEOUT);

i will suggest you to reduce the "TIMEOUT"  constant instead of hard coding it to "1" second.

djelinski replied:
> I can see that this test uses "TIMEOUT" down in test

Right, this is why I didn't remove the TIMEOUT constant
> i will suggest you to reduce the "TIMEOUT" constant instead of hard coding it to "1" second.

I can't; this would make the test fail more often. I could introduce a separate timeout constant for `handshakeTimeout` to make it clear that the two are not related.

Note that this test uses 2 timeouts:
- the `handshakeTimeout`, which is the time limit for socket operations; when this amount of time passes waiting for a response from the server, an exception is thrown
- the `TIMEOUT`, which is the amount of time after which we expect to get the exception. This includes the `handshakeTimeout`, but also includes all other operations performed during the handshake, like class loading, initialization, TLS handshake etc.

Alternatively I could change the test descriptor to:
```
 * @run main/othervm/timeout=10 -Dsun.rmi.transport.tcp.handshakeTimeout=1 HandshakeTimeout
```
and drop both constants from the code.
What do you think?

vyommani replied:
setting System property(sun.rmi.transport.tcp.handshakeTimeout) as argument is good idea and this will increase the code readability as well.

msheppar commented:
What is the motivation for reducing the timeout for the handshake? 

I think the failure in the bug hints at a few potential issues that may be worth a deeper look.

There are two runs of this test ... plain socket and ssl socket. The failures is on the second ssl scenario. 

This is unexpected and question arises why did it happen. Is it an obscure issue in the RMI layer, or in the SSL layer, or perhaps in the native code associated with "select" for the timeout on the read/write ? 

Reducing the handshake timeout for the first scenario seems ok. But reducing the handshake to 1 millisecond for the ssl scenario maybe masking an issue with the SSL timers or within the RMI layer -- the failure could be suggesting some deeper issues. 

What the failure suggestion is that additional diagnostics would be helpful for determining how the main thread has reached the no timeout condition: 
         if (connector.exception == null) { 
                throw new RuntimeException( 
                    "TEST FAILED: remote call did not time out"); 
            } else { 

Did the Connector thread actually run? Is it in the RTR state waiting to be scheduled when the test completed? is it still alive? should additional state be added to indicate that the Connector thread has started? Should the Connector thread and the main thread synchronise on a countDownLatch before the thread join ? 

Thus, the failure suggests that a bit of restructuring would be useful, to add a bit of Connect run state, or check the state of the Connector thread prior to join, and some synchronization between the Connector thread and the main thread before the join to ensure that the Connector thread has started. The main thread waits for the Connector thread to actually start.

djelinski commented:
The original motivation to touch this test was that the SSL version timed out; the handshake timeout was too long in proportion to the test timeout, which was fixed at 10 seconds (see `TIMEOUT` constant).
The `sun.rmi.transport.tcp.handshakeTimeout` property only applies to a single network operation in the current implementation. SSL part of the handshake (client hello/key share generation) does not count towards the timeout, and if there are multiple network operations (e.g. when doing SSL handshake we first write to the socket, then read from it), each operation can use the full timeout value.

The purpose of this test is to verify that the handshake does not block forever (JDK-4322806 was for plain handshake, JDK-8189338 was for SSL handshake). After my changes the test still does that; without the fixes for the above bugs, the test times out.

Thanks for the restructuring suggestion. I modified the test to run in a single thread. Test timeout is now handled by JTReg timeout handler, which means that 1) we will have a thread dump when that happens, 2) timeout factor will be used in timeout calculations.

Let me know if that addresses your concerns.

msheppar commented:
the restructuring is good idea. I never think of running clint and server in same thread, but it will eliminate any possibility of race between main thread and Connector thread. so job done (LGTM)

vyommani reviewed (approved):
Refactored test looks good to me.

djelinski commented:
/integrate
